pkg/operator/controller/dnsnameresolver/resolver.go

```diff
@@ -0,0 +1,340 @@
+package dnsnameresolver
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/miekg/dns"
+	networkv1alpha1 "github.com/openshift/api/network/v1alpha1"
+	"github.com/sirupsen/logrus"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	defaultMaxTTL    = 30 * time.Minute
+	defaultMinTTL    = 5 * time.Second
+	defaultPort      = "53"
+	maxCoreDNSPodIPs = 5
+)
+
+// dnsResolvedName contains details about a DNS resolved name.
+type dnsResolvedName struct {
+	minNextLookupTime time.Time
+	regularObjExists  bool
+	wildcardObjExists bool
+	numIPs            int
+}
+
+// Resolver keeps track of the DNS names, the corresponding IP addresses and TTLs. It sends DNS resolution
+// requests to CoreDNS pods whenever TTL of any IP address associated to a DNS name expires.
```

Comment:
Isn't the TTL associated with the DNS name,  and not the IP address?  Are we really supposed to be checking every IP address for TTL?

Reply:
This was discussed extensively during the EP phase and it was decided to track the TTL of each IP address: https://github.com/openshift/enhancements/pull/1335#discussion_r1280694158

Reply:
Moved to openshift/coredns-ocp-dnsnameresolver#2